crates/prover/src/adapter/instructions.rs

```diff
@@ -0,0 +1,474 @@
+    pub store_deref_fp: Vec<StateData>,  // [fp + off2] = [fp + off0]
+    pub store_double_deref_fp: Vec<StateData>, // [fp + off2] = [[fp + off0] + off1]
+    pub store_imm: Vec<StateData>,       // [fp + off2] = imm
+    pub store_mul_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] * [fp + off1]
+    pub store_mul_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] * imm
+    pub store_div_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] / [fp + off1]
+    pub store_div_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] / imm
+    pub call_abs_fp: Vec<StateData>,     // call abs [fp + off0]
+    pub call_abs_imm: Vec<StateData>,    // call abs imm
+    pub call_rel_fp: Vec<StateData>,     // call rel [fp + off0]
+    pub call_rel_imm: Vec<StateData>,    // call rel imm
+    pub ret: Vec<StateData>,             // ret
+    pub jmp_abs_add_fp_fp: Vec<StateData>, // jmp abs [fp + off0] + [fp + off1]
+    pub jmp_abs_add_fp_imm: Vec<StateData>, // jmp abs [fp + off0] + imm
+    pub jmp_abs_deref_fp: Vec<StateData>, // jmp abs [fp + off0]
+    pub jmp_abs_double_deref_fp: Vec<StateData>, // jmp abs [[fp + off0] + off1]
+    pub jmp_abs_imm: Vec<StateData>,     // jmp abs imm
+    pub jmp_abs_mul_fp_fp: Vec<StateData>, // jmp abs [fp + off0] * [fp + off1]
+    pub jmp_abs_mul_fp_imm: Vec<StateData>, // jmp abs [fp + off0] * imm
+    pub jmp_rel_add_fp_fp: Vec<StateData>, // jmp rel [fp + off0] + [fp + off1]
+    pub jmp_rel_add_fp_imm: Vec<StateData>, // jmp rel [fp + off0] + imm
+    pub jmp_rel_deref_fp: Vec<StateData>, // jmp rel [fp + off0]
+    pub jmp_rel_double_deref_fp: Vec<StateData>, // jmp rel [[fp + off0] + off1]
+    pub jmp_rel_imm: Vec<StateData>,     // jmp rel imm
+    pub jmp_rel_mul_fp_fp: Vec<StateData>, // jmp rel [fp + off0] * [fp + off1]
+    pub jmp_rel_mul_fp_imm: Vec<StateData>, // jmp rel [fp + off0] * imm
+    pub jnz_fp_fp: Vec<StateData>,       // jmp rel [fp + off1] if [fp + off0] != 0
+    pub jnz_fp_imm: Vec<StateData>,
+}
+
+#[derive(Debug, Default)]
+pub struct Instructions {
+    pub initial_registers: VmRegisters,
+    pub final_registers: VmRegisters,
+    pub states_by_opcodes: StatesByOpcodes,
+}
+
+impl Instructions {
+    #[allow(clippy::cognitive_complexity)]
+    pub fn push_instr<I>(
+        &mut self,
+        mut memory: I,
+        registers: VmRegisters,
+        clock: u32,
+        memory_cache: &mut MemoryCache,
+    ) -> Result<(), InstructionError>
+    where
+        I: Iterator<Item = MemoryEntry>,
+    {
+        let mut state_data = StateData {
+            registers,
+            memory_args: Default::default(),
+        };
+
+        let instruction = memory
+            .next()
+            .ok_or(InstructionError::UnexpectedEndOfMemory)?;
+        let opcode_id = Opcode::from(instruction.value).id();
+        state_data.memory_args[0] = memory_cache.push(instruction, clock);
+
+        match opcode_id {
```

Comment:
with the `opcode` enum implemented as above, you can remove the whole match and just gets the number of items to "next" from the memory like `memory.take(n).collect()`